docs/slices.md

```diff
@@ -0,0 +1,1881 @@
+======
+
+Python's slice syntax is one of the more confusing parts of the language, even
+to experienced developers. In this page, I carefully break down the rules for
+slicing, and examine just what it is that makes it so confusing.
+
+There are two primary aspects of slices that make them difficult to
+understand: confusing conventions, and discontinuous definitions. By confusing
+conventions, we mean that slice semantics have definitions that are often
+difficult to reason about mathematically. These conventions were chosen for
+syntactic convenience, and one can easily see for most of them how they lead
+to concise notation for very common operations, but it remains nonetheless
+true that they can make figuring out the *right* slice to use in the first
+place complicated. By discontinuous definitions, we mean that the definition
+of a slice takes on fundamentally different meanings if the start, stop, or
+step are negative, nonnegative, or omitted. This again is done for syntactic
+convenience, but it means that as a user, you must switch your mode of
+thinking about slices depending on value of the arguments. There are no
+uniform formulas that apply to all slices.
+
+The [ndindex](index) library can help with much of this, especially for people
+developing libraries that consume slices. But for end-users the challenge is
+often just to write down a slice. Even if you rarely work with NumPy arrays,
+you will most likely require slices to select parts of lists or strings as
+part of the normal course of Python coding.
+
+ndindex focuses on NumPy array index semantics, but everything on this page
+equally applies to sliceable Python builtin objects like lists, tuples, and
+strings. This is because on a single dimension, NumPy slice semantics are
+identical to the Python slice semantics (NumPy only begins to differ from
+Python for multi-dimensional indices).
+
+## What is a slice?
+
+In Python, a slice is a special syntax that is allowed only in an index, that
+is, inside of square brackets proceeding an expression. A slice consists of
+one or two colons, with either an expression or nothing on either side of each
+colon. For example, the following are all valid slices on the object `a`:
+
+```py
+a[x:y]
+a[x:y:z]
+a[:]
+a[x::]
+a[x::z]
+```
+
+Furthermore, for a slice `a[x:y:z]`, if `a` is a Python built-in object or a
+NumPy array, there is an additional semantic restriction, which is that the
+expressions `x`, `y`, and `z` must be integers.
+
+The three arguments to a slice are traditionally called `start`, `stop`, and
+`step`:
+
+```py
+a[start:stop:step]
+```
+
+We will use these names throughout this guide.
+
+It is worth noting that the `x:y:z` syntax is not valid outside of square
+brackets. However, slice objects can be created manually using the `slice()`
+builtin (`a[x:y:z]` is the same as `a[slice(x, y, z)]`). You can also use the
+[`ndindex.Slice()`](slice-api) object if you want to perform more advanced
+operations.
+
+(rules)=
```

Comment:
This is rendering kind of oddly in the github preview

Reply:
Yeah, this is MyST specific syntax for cross-references, so other Markdown renderers won't know about it and will just render it literally.